File: core/RotateArray/RotateArray.py
from typing import List


def rotate(nums: List[int], k: int) -> None:
    nums.reverse()
    k = (k - 1) % len(nums)
    left, right = 0, k
    while left < right:
        nums[left], nums[right] = nums[right], nums[left]
        left += 1
        right -= 1
    left, right = k + 1, len(nums) - 1
    while left < right:
        nums[left], nums[right] = nums[right], nums[left]
        left += 1
        right -= 1

File: core/RotateArray/RotateArray.java
package com.dbc.code;

import java.util.Arrays;
import java.util.Collection;

public class RotateArray {
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }

    public void rotate(int[] nums, int k) {
        reverse(nums, 0, nums.length - 1);
        reverse(nums, 0, (k - 1) % nums.length);
        reverse(nums, (k - 1) % nums.length + 1, nums.length - 1);
    }
}
